hilo.go

```diff
@@ -0,0 +1,271 @@
+		URLTemplate: url,
+	}
+	return res
+}
+
+// NextHiLoResult is a result of NextHiLoResult command
+type NextHiLoResult struct {
+	Prefix      string `json:"Prefix"`
+	Low         int    `json:"Low"`
+	High        int    `json:"High"`
+	LastSize    int    `json:"LastSize"`
+	ServerTag   string `json:"ServerTag"`
+	LastRangeAt string `json:"LastRangeAt"`
+}
+
+const (
+	// time format returned by the server
+	// 2018-05-08T05:20:31.5233900Z
+	serverTimeFormat = "2006-01-02T15:04:05.999999999Z"
+)
+
+// GetLastRangeAt parses LastRangeAt which is in a format:
+// 2018-05-08T05:20:31.5233900Z
+func (r *NextHiLoResult) GetLastRangeAt() time.Time {
+	t, err := time.Parse(serverTimeFormat, r.LastRangeAt)
+	must(err) // TODO: should silently fail? return an error?
+	return t
+}
+
+// ExecuteNewNextHiLoCommand executes NextHiLoResult command
+func ExecuteNewNextHiLoCommand(exec CommandExecutorFunc, cmd *RavenCommand) (*NextHiLoResult, error) {
+	var res NextHiLoResult
+	err := excuteCmdAndJSONDecode(exec, cmd, &res)
+	if err != nil {
+		return nil, err
+	}
+	return &res, nil
+}
+
+// HiLoKeyGenerator generates keys server side
+// https://sourcegraph.com/github.com/ravendb/RavenDB-Python-Client@v4.0/-/blob/pyravendb/hilo/hilo_generator.py#L63
+// https://sourcegraph.com/github.com/ravendb/ravendb-jvm-client@v4.0/-/blob/src/main/java/net/ravendb/client/documents/identity/HiLoIdGenerator.java#L14
+type HiLoKeyGenerator struct {
+	tag                    string
+	store                  *DocumentStore
+	dbName                 string
+	lastRangeAt            time.Time
+	lastBatchSize          int
+	rangev                 *RangeValue
+	prefix                 string
+	serverTag              string
+	convetions             *DocumentConventions
+	identityPartsSeparator string
+	lock                   sync.Mutex
+}
+
+// NewHiLoKeyGenerator creates a HiLoKeyGenerator
+func NewHiLoKeyGenerator(tag string, store *DocumentStore, dbName string) *HiLoKeyGenerator {
+	if dbName == "" {
+		dbName = store.database
+	}
+	res := &HiLoKeyGenerator{
+		tag:           tag,
+		store:         store,
+		dbName:        dbName,
+		lastRangeAt:   time.Date(1, 1, 1, 0, 0, 0, 0, time.UTC),
+		lastBatchSize: 0,
+		rangev:        NewRangeValue(1, 0),
+		prefix:        "",
+		serverTag:     "",
+		convetions:    store.Conventions,
+	}
+	res.identityPartsSeparator = res.convetions.IdentityPartsSeparator
+	return res
+}
+
+// GetDocumentKeyFromID creates key from id
+func (g *HiLoKeyGenerator) GetDocumentKeyFromID(nextID int) string {
+	return fmt.Sprintf("%s%d-%s", g.prefix, nextID, g.serverTag)
+}
+
+// GenerateDocumentKey returns next key
+func (g *HiLoKeyGenerator) GenerateDocumentKey() string {
+	// TODO: propagate error
+	id, _ := g.nextID()
+	return g.GetDocumentKeyFromID(id)
+}
+
+func (g *HiLoKeyGenerator) nextID() (int, error) {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+	for {
+		// local range is not exhausted yet
+		rangev := g.rangev
+		id := rangev.Next()
```

Comment:
We can reduce the lock scope significantly here, if we have `Next` that is atomic

Reply:
Added a TODO for that.